cpp/src/io/json/experimental/read_json.cpp

```diff
@@ -38,29 +39,48 @@ size_t sources_size(host_span<std::unique_ptr<datasource>> const sources,
   });
 }
 
-std::vector<uint8_t> ingest_raw_input(host_span<std::unique_ptr<datasource>> const& sources,
-                                      compression_type compression,
-                                      size_t range_offset,
-                                      size_t range_size)
+rmm::device_uvector<char> ingest_raw_input(host_span<std::unique_ptr<datasource>> const& sources,
+                                           compression_type compression,
+                                           size_t range_offset,
+                                           size_t range_size,
+                                           rmm::cuda_stream_view stream)
 {
   CUDF_FUNC_RANGE();
   // Iterate through the user defined sources and read the contents into the local buffer
   auto const total_source_size = sources_size(sources, range_offset, range_size);
-  auto buffer                  = std::vector<uint8_t>(total_source_size);
-
-  size_t bytes_read = 0;
-  for (const auto& source : sources) {
-    if (!source->is_empty()) {
-      auto data_size   = (range_size != 0) ? range_size : source->size();
-      auto destination = buffer.data() + bytes_read;
-      bytes_read += source->host_read(range_offset, data_size, destination);
-    }
-  }
 
   if (compression == compression_type::NONE) {
-    return buffer;
+    auto d_buffer     = rmm::device_uvector<char>(total_source_size, stream);
+    size_t bytes_read = 0;
+    for (const auto& source : sources) {
+      if (!source->is_empty()) {
+        auto data_size   = (range_size != 0) ? range_size : source->size();
+        auto destination = reinterpret_cast<uint8_t*>(d_buffer.data()) + bytes_read;
+        if (source->is_device_read_preferred(data_size)) {
+          bytes_read += source->device_read(range_offset, data_size, destination, stream);
+        } else {
+          auto const h_buffer = source->host_read(range_offset, data_size);
+          CUDF_CUDA_TRY(cudaMemcpyAsync(destination,
+                                        h_buffer->data(),
+                                        h_buffer->size(),
+                                        cudaMemcpyHostToDevice,
```

Comment:
Replaced across the codebase in #12466.
```suggestion
                                        cudaMemcpyDefault,
```